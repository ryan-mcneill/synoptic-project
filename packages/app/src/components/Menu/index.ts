import Accordion from "./Accordion";
import InfoTile from "./InfoTile";
import PlaylistTab from "./PlaylistTab";
import SearchTab from "./SearchTab";
import SongList from "./SongList";
import SongMenu from "./SongMenu";
import SongsTab from "./SongsTab";
import TabButton from "./TabButton";
import Tabs from "./Tabs";

export {
  Accordion,
  InfoTile,
  PlaylistTab,
  SearchTab,
  SongList,
  SongMenu,
  SongsTab,
  TabButton,
  Tabs
};
