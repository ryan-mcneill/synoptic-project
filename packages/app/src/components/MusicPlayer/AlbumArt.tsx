import React, { ReactElement } from "react";
import { connect, ConnectedProps } from "react-redux";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faMusic } from "@fortawesome/free-solid-svg-icons";
import { State } from "../../store/types";

const mapStateToProps = (state: State) => ({
  albumId: state.playlist.current?.albums[0]._id
});

const connector = connect(mapStateToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

const AlbumArt: React.FC<PropsFromRedux> = ({ albumId }): ReactElement => {
  return (
    <div
      className="album-art"
      style={{
        width: "80vw",
        height: "80vw",
        position: "absolute",
        top: "42vh",
        left: "50vw",
        transform: "translate(-50%, -50%)",
        backgroundColor: "#c0c0c0",
        boxShadow: "0px 0px 64px 4px rgba(0,0,0,0.2)"
      }}
    >
      {albumId ? (
        <img
          alt="Album art"
          src={`/api/album/art/${albumId}`}
          width="100%"
          height="100%"
        />
      ) : (
        <div
          style={{
            fontSize: "400px",
            position: "absolute",
            top: "50%",
            left: "48%",
            transform: "translate(-50%, -50%)"
          }}
        >
          <FontAwesomeIcon icon={faMusic} />
        </div>
      )}
    </div>
  );
};

export default connector(AlbumArt);
