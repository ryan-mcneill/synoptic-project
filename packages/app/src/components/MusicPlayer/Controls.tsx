import React, { ReactElement, useState } from "react";
import { connect, ConnectedProps } from "react-redux";
import classnames from "classnames";
import ReactAudioPlayer from "react-audio-player";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faPause,
  faPlay,
  faStepBackward,
  faStepForward
} from "@fortawesome/free-solid-svg-icons";
import { nextSong, prevSong, togglePlaying } from "../../store/actions";
import { State } from "../../store/types";

const mapDispatchToProps = (dispatch: any) => ({
  togglePlaying: (playing?: boolean) => dispatch(togglePlaying(playing)),
  nextSong: () => dispatch(nextSong()),
  prevSong: () => dispatch(prevSong())
});

const mapStateToProps = (state: State) => ({
  isDark: state.currentSong.background?.isDark,
  isPlaying: state.currentSong?.isPlaying,
  songId: state.playlist.current?._id,
  data: state.playlist.current
});

const connector = connect(mapStateToProps, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

const Controls: React.FC<PropsFromRedux> = ({
  data,
  isDark,
  isPlaying,
  songId,
  togglePlaying,
  nextSong,
  prevSong
}): ReactElement => {
  const [songLength, setSongLength] = useState(0);
  const [songTime, setSongTime] = useState(0);
  let audioPlayer: ReactAudioPlayer;

  const onSongLoad = () => {
    if (audioPlayer?.audioEl?.current) {
      if (isPlaying) audioPlayer.audioEl.current.play();
      setSongLength(audioPlayer.audioEl.current.duration);
    }
  };

  const formatTime = (time: number): string => {
    const tempDate = new Date(Math.round(time) * 1000);
    const minutes = tempDate.getMinutes();
    const seconds = tempDate.getSeconds();

    return (
      minutes.toString().padStart(2, "0") +
      ":" +
      seconds.toString().padStart(2, "0")
    );
  };

  const playStatusHandler = () => {
    if (data && audioPlayer?.audioEl?.current) {
      togglePlaying();
      audioPlayer.audioEl.current[isPlaying ? "pause" : "play"]();
    }
  };

  const nextHandler = () => {
    if (data) {
      nextSong();
      setSongTime(0);
    }
  };

  const prevHandler = () => {
    if (data) {
      prevSong();
      setSongTime(0);
    }
  };

  return (
    <div
      style={{
        height: "10vh",
        width: "90vw",
        position: "absolute",
        bottom: 0,
        padding: "0 5vw 5vh 5vw"
      }}
    >
      <ReactAudioPlayer
        ref={(element: ReactAudioPlayer) => (audioPlayer = element)}
        src={songId ? `/api/song/${songId}` : ""}
        listenInterval={1000}
        onListen={(e) => setSongTime(e)}
        onLoadedMetadata={() => onSongLoad()}
        onEnded={() => nextHandler()}
      />
      <div
        style={{
          display: "flex",
          alignItems: "center",
          justifyContent: "center"
        }}
      >
        <div style={{ fontSize: "48px" }} onClick={() => prevHandler()}>
          <FontAwesomeIcon icon={faStepBackward} />
        </div>
        <div
          style={{ fontSize: "96px", margin: "0 96px" }}
          onClick={playStatusHandler}
        >
          <FontAwesomeIcon icon={isPlaying ? faPause : faPlay} />
        </div>
        <div style={{ fontSize: "48px" }} onClick={() => nextHandler()}>
          <FontAwesomeIcon icon={faStepForward} />
        </div>
      </div>
      <div
        style={{
          display: "flex",
          alignItems: "center",
          justifyContent: "center",
          marginTop: "48px",
          fontSize: "32px"
        }}
      >
        <div>{formatTime(songTime)}</div>
        <input
          className={classnames({ dark: !isDark, light: isDark })}
          type="range"
          style={{ width: "100%", margin: "0 24px" }}
          min="0"
          max={songLength}
          defaultValue={songTime}
        />
        <div>{formatTime(songLength)}</div>
      </div>
    </div>
  );
};

export default connector(Controls);
