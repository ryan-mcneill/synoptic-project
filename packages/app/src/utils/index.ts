import { artistsToString } from "./artistsToString";
import { filterData } from "./filterData";
import { filterSongs } from "./filterSongs";
import { playlistsToArray } from "./playlistsToArray";

export { artistsToString, filterData, filterSongs, playlistsToArray };
