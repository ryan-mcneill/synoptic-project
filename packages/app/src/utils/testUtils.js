import React from "react";
import { render as rtlRender } from "@testing-library/react";
import { createStore } from "redux";
import { Provider } from "react-redux";
import reducers from "../store/reducers";

function render(
  ui,
  {
    initialState,
    store = createStore(reducers, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from "@testing-library/react";
export { render };
