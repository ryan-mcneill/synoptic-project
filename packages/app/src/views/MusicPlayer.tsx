import React, { ReactElement } from "react";
import { connect, ConnectedProps } from "react-redux";
import { AlbumArt, Controls, NowPlaying } from "../components/MusicPlayer";
import { State } from "../store/types";

const mapStateToProps = (state: State) => ({
  albumsLoading: state.albums.loading,
  artistsLoading: state.artists.loading,
  isDark: state.currentSong.background?.isDark,
  backgroundColor: state.currentSong.background?.colour
});

const connector = connect(mapStateToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

const MusicPlayer: React.FC<PropsFromRedux> = ({
  albumsLoading,
  artistsLoading,
  isDark,
  backgroundColor
}): ReactElement =>
  !albumsLoading && !artistsLoading ? (
    <div
      style={{
        display: "block",
        margin: 0,
        position: "absolute",
        height: "100%",
        width: "100%",
        color: isDark ? "white" : "black",
        fontSize: "18px",
        backgroundColor
      }}
    >
      <NowPlaying />
      <AlbumArt />
      <Controls />
    </div>
  ) : (
    <></>
  );

export default connector(MusicPlayer);
