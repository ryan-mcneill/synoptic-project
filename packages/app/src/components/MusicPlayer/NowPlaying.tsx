import React, { ReactElement } from "react";
import { connect, ConnectedProps } from "react-redux";
import { artistsToString } from "../../utils";
import { State } from "../../store/types";

const mapStateToProps = (state: State) => ({
  artists: state.playlist.current?.artists,
  songName: state.playlist.current?.name,
  data: state.playlist.current
});

const connector = connect(mapStateToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

const NowPlaying: React.FC<PropsFromRedux> = ({
  artists = [{ name: "Unknown Artist" }],
  songName = "Unknown",
  data
}): ReactElement => {
  return (
    <div
      style={{
        height: "5vh",
        width: "90vw",
        textAlign: "center",
        padding: "5vh 5vw 2vh 5vw"
      }}
    >
      <div style={{ fontSize: "24px" }}>
        {data ? "NOW PLAYING" : "NOTHING PLAYING"}
      </div>
      <div
        style={{
          fontSize: "48px",
          whiteSpace: "nowrap",
          overflow: "hidden",
          textOverflow: "ellipsis"
        }}
      >
        {data
          ? `${songName} - ${artistsToString(artists)}`
          : "Please make a song selection"}
      </div>
    </div>
  );
};

export default connector(NowPlaying);
