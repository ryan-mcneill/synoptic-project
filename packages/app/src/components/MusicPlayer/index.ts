import AlbumArt from "./AlbumArt";
import Controls from "./Controls";
import NowPlaying from "./NowPlaying";

export { AlbumArt, Controls, NowPlaying };
