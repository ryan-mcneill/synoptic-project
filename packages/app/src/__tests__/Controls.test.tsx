import React from "react";
import { Controls } from "../components/MusicPlayer";
import { render } from "../utils/testUtils";

describe("The Controls component", () => {
  it("renders correctly with default props", () => {
    const { container } = render(<Controls />, {
      initialState: { currentSong: { isPlaying: false } }
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders correctly on a dark background", () => {
    const { container } = render(<Controls />, {
      initialState: {
        currentSong: { isPlaying: false, background: { isDark: true } }
      }
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders with the data loaded and the correct song playing", () => {
    const { container } = render(<Controls />, {
      initialState: {
        currentSong: { isPlaying: false },
        playlist: { current: { _id: "1234567890" }}
      }
    });

    expect(container.firstChild).toMatchSnapshot();
  });
});
