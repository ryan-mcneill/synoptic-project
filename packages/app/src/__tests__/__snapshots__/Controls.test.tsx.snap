// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Controls component renders correctly on a dark background 1`] = `
<div
  style="height: 10vh; width: 90vw; position: absolute; bottom: 0px; padding: 0px 5vw 5vh 5vw;"
>
  <audio
    class="react-audio-player "
    id=""
    preload="metadata"
    src=""
    title=""
  >
    <p>
      Your browser does not support the 
      <code>
        audio
      </code>
       element.
    </p>
  </audio>
  <div
    style="display: flex; align-items: center; justify-content: center;"
  >
    <div
      style="font-size: 48px;"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-step-backward fa-w-14 "
        data-icon="step-backward"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M64 468V44c0-6.6 5.4-12 12-12h48c6.6 0 12 5.4 12 12v176.4l195.5-181C352.1 22.3 384 36.6 384 64v384c0 27.4-31.9 41.7-52.5 24.6L136 292.7V468c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      style="font-size: 96px; margin: 0px 96px;"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-play fa-w-14 "
        data-icon="play"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      style="font-size: 48px;"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-step-forward fa-w-14 "
        data-icon="step-forward"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M384 44v424c0 6.6-5.4 12-12 12h-48c-6.6 0-12-5.4-12-12V291.6l-195.5 181C95.9 489.7 64 475.4 64 448V64c0-27.4 31.9-41.7 52.5-24.6L312 219.3V44c0-6.6 5.4-12 12-12h48c6.6 0 12 5.4 12 12z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
  <div
    style="display: flex; align-items: center; justify-content: center; margin-top: 48px; font-size: 32px;"
  >
    <div>
      00:00
    </div>
    <input
      class="light"
      max="0"
      min="0"
      style="width: 100%; margin: 0px 24px;"
      type="range"
      value="0"
    />
    <div>
      00:00
    </div>
  </div>
</div>
`;

exports[`The Controls component renders correctly with default props 1`] = `
<div
  style="height: 10vh; width: 90vw; position: absolute; bottom: 0px; padding: 0px 5vw 5vh 5vw;"
>
  <audio
    class="react-audio-player "
    id=""
    preload="metadata"
    src=""
    title=""
  >
    <p>
      Your browser does not support the 
      <code>
        audio
      </code>
       element.
    </p>
  </audio>
  <div
    style="display: flex; align-items: center; justify-content: center;"
  >
    <div
      style="font-size: 48px;"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-step-backward fa-w-14 "
        data-icon="step-backward"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M64 468V44c0-6.6 5.4-12 12-12h48c6.6 0 12 5.4 12 12v176.4l195.5-181C352.1 22.3 384 36.6 384 64v384c0 27.4-31.9 41.7-52.5 24.6L136 292.7V468c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      style="font-size: 96px; margin: 0px 96px;"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-play fa-w-14 "
        data-icon="play"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      style="font-size: 48px;"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-step-forward fa-w-14 "
        data-icon="step-forward"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M384 44v424c0 6.6-5.4 12-12 12h-48c-6.6 0-12-5.4-12-12V291.6l-195.5 181C95.9 489.7 64 475.4 64 448V64c0-27.4 31.9-41.7 52.5-24.6L312 219.3V44c0-6.6 5.4-12 12-12h48c6.6 0 12 5.4 12 12z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
  <div
    style="display: flex; align-items: center; justify-content: center; margin-top: 48px; font-size: 32px;"
  >
    <div>
      00:00
    </div>
    <input
      class="dark"
      max="0"
      min="0"
      style="width: 100%; margin: 0px 24px;"
      type="range"
      value="0"
    />
    <div>
      00:00
    </div>
  </div>
</div>
`;

exports[`The Controls component renders with the data loaded and the correct song playing 1`] = `
<div
  style="height: 10vh; width: 90vw; position: absolute; bottom: 0px; padding: 0px 5vw 5vh 5vw;"
>
  <audio
    class="react-audio-player "
    id=""
    preload="metadata"
    src="http://localhost:5000/api/song/1234567890"
    title="http://localhost:5000/api/song/1234567890"
  >
    <p>
      Your browser does not support the 
      <code>
        audio
      </code>
       element.
    </p>
  </audio>
  <div
    style="display: flex; align-items: center; justify-content: center;"
  >
    <div
      style="font-size: 48px;"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-step-backward fa-w-14 "
        data-icon="step-backward"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M64 468V44c0-6.6 5.4-12 12-12h48c6.6 0 12 5.4 12 12v176.4l195.5-181C352.1 22.3 384 36.6 384 64v384c0 27.4-31.9 41.7-52.5 24.6L136 292.7V468c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      style="font-size: 96px; margin: 0px 96px;"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-play fa-w-14 "
        data-icon="play"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      style="font-size: 48px;"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-step-forward fa-w-14 "
        data-icon="step-forward"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M384 44v424c0 6.6-5.4 12-12 12h-48c-6.6 0-12-5.4-12-12V291.6l-195.5 181C95.9 489.7 64 475.4 64 448V64c0-27.4 31.9-41.7 52.5-24.6L312 219.3V44c0-6.6 5.4-12 12-12h48c6.6 0 12 5.4 12 12z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
  <div
    style="display: flex; align-items: center; justify-content: center; margin-top: 48px; font-size: 32px;"
  >
    <div>
      00:00
    </div>
    <input
      class="dark"
      max="0"
      min="0"
      style="width: 100%; margin: 0px 24px;"
      type="range"
      value="0"
    />
    <div>
      00:00
    </div>
  </div>
</div>
`;
